File: packages/core/src/utils/task-sort.ts
import type { Task } from '@track/shared'
import { PRIORITY_RANK, STATUS_RANK } from '@track/shared'

export function sortTasks(tasks: Task[]): Task[] {
  return [...tasks].sort((left, right) => {
    const statusDifference = STATUS_RANK[left.status] - STATUS_RANK[right.status]
    if (statusDifference !== 0) {
      return statusDifference
    }

    const priorityDifference = PRIORITY_RANK[left.priority] - PRIORITY_RANK[right.priority]
    if (priorityDifference !== 0) {
      return priorityDifference
    }

    return right.createdAt.localeCompare(left.createdAt)
  })
}

File: crates/track-core/src/task_sort.rs
use crate::types::Task;

pub fn sort_tasks(tasks: &[Task]) -> Vec<Task> {
    let mut sorted = tasks.to_vec();
    sorted.sort_by(|left, right| {
        left.status
            .cmp(&right.status)
            .then_with(|| left.priority.cmp(&right.priority))
            .then_with(|| right.created_at.cmp(&left.created_at))
    });
    sorted
}
